applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_adjoint_static_solver.py

```diff
@@ -0,0 +1,127 @@
+        KratosMultiphysics.VariableUtils().AddDof(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT_Y, self.main_model_part)
+        KratosMultiphysics.VariableUtils().AddDof(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT_Z, self.main_model_part)
+        if self.settings["rotation_dofs"].GetBool():
+            KratosMultiphysics.VariableUtils().AddDof(StructuralMechanicsApplication.ADJOINT_ROTATION_X, self.main_model_part)
+            KratosMultiphysics.VariableUtils().AddDof(StructuralMechanicsApplication.ADJOINT_ROTATION_Y, self.main_model_part)
+            KratosMultiphysics.VariableUtils().AddDof(StructuralMechanicsApplication.ADJOINT_ROTATION_Z, self.main_model_part)
+        self.print_on_rank_zero("::[AdjointMechanicalSolver]:: ", "DOF's ADDED.")
+
+    def Initialize(self):
+        """Perform initialization after adding nodal variables and dofs to the main model part. """
+        if self.response_function_settings["response_type"].GetString() == "adjoint_local_stress":
+            self.response_function = StructuralMechanicsApplication.AdjointLocalStressResponseFunction(self.main_model_part, self.response_function_settings)
+        elif self.response_function_settings["response_type"].GetString() == "adjoint_nodal_displacement":
+            self.response_function = StructuralMechanicsApplication.AdjointNodalDisplacementResponseFunction(self.main_model_part, self.response_function_settings)
+        elif self.response_function_settings["response_type"].GetString() == "adjoint_strain_energy":
+            self.response_function = StructuralMechanicsApplication.AdjointStrainEnergyResponseFunction(self.main_model_part, self.response_function_settings)
+        else:
+            raise Exception("invalid response_type: " + self.response_function_settings["response_type"].GetString())
+
+        super(StructuralMechanicsAdjointStaticSolver, self).Initialize()
+
+        self.print_on_rank_zero("::[AdjointMechanicalSolver]:: ", "Finished initialization.")
+
+    def Solve(self):
+        if self.response_function_settings["response_type"].GetString() == "adjoint_strain_energy":
+            self._SolveSolutionStepSpecialStrainEnergy()
+        else:
+            if self.settings["clear_storage"].GetBool():
+                self.Clear()
+            mechanical_solution_strategy = self.get_mechanical_solution_strategy()
+            mechanical_solution_strategy.Solve()
+
+    def SolveSolutionStep(self):
+        if self.response_function_settings["response_type"].GetString() == "adjoint_strain_energy":
+            self._SolveSolutionStepSpecialStrainEnergy()
+        else:
+            super(StructuralMechanicsAdjointStaticSolver, self).SolveSolutionStep()
+
+    def _SolveSolutionStepSpecialStrainEnergy(self):
+        self.response_function.Initialize()
+        for node in self.main_model_part.Nodes:
```

Comment:
There is a function in the variable utils for this

Reply:
I can't find a function that copies and multiplies at the same time. So it remains essentially the same, i have to loop in python at least once with get/set because we need to multiply everything with `0.5`